DOC Updates for symfony/mailer

emteknetnz opened the pull request:
Issue https://github.com/silverstripe/silverstripe-framework/issues/10097

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
 # Email
 
-Creating and sending email in Silverstripe CMS is done through the [Email](api:SilverStripe\Control\Email\Email) and [Mailer](api:SilverStripe\Control\Email\Mailer) classes. This document covers how to create an `Email` instance, customise it with a HTML template, then send it through a custom `Mailer`.
+Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email). This document covers how to create an `Email` instance and customise it with an HTML template.
```
```suggestion
Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email) class. This document covers how to create an `Email` instance and customise it with an HTML template.
```

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
 ## Configuration
 
-Silverstripe CMS provides an API over the top of the [SwiftMailer](http://swiftmailer.org/) PHP library which comes with an extensive list of "transports" for sending mail via different services. 
+Silverstripe CMS provides an API over the top of [symfony/mailer](https://symfony.com/doc/current/mailer.html) which comes with an extensive list of "transports" for sending mail via different services such as SMTP, Gmail and Amazon SES.
```
```suggestion
Silverstripe CMS provides an API over the top of [symfony/mailer](https://symfony.com/doc/current/mailer.html) which comes with an extensive list of "transports" for sending mail via different services such as SMTP, Gmail, and Amazon SES.
```

emteknetnz replied:
We could argue all day about this https://www.grammarly.com/blog/comma-before-and/

ChloeSartorelli replied:
Does that mean you'll make the change to avoid an argument or you won't make the change and I'll just have to live without my oxford comma?

emteknetnz replied:
I performed a string search for 
`, and`
in 
`vendor/silverstripe/**/*.md`

There are numerous example of using a comma before `and` in a list, much to my horror :-)

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
+Silverstripe CMS provides an API over the top of [symfony/mailer](https://symfony.com/doc/current/mailer.html) which comes with an extensive list of "transports" for sending mail via different services such as SMTP, Gmail and Amazon SES.
 
-For legacy reasons, Silverstripe CMS defaults to using the built-in PHP `mail()` command via a deprecated class `Swift_MailTransport`. However, using this layer is less secure and is strongly discouraged.
+Email configuration is done using Symfony's `[DSN](https://symfony.com/doc/current/mailer.html#transport-setup) configuration string which is used to select which transport is used and any required configuration such as username and password. In Silverstripe, this is done with either an environment variable or yml configuration.
```
```suggestion
Email configuration is done using Symfony's [DSN](https://symfony.com/doc/current/mailer.html#transport-setup) configuration string which is used to select which transport is used and any required configuration such as username and password. In Silverstripe, this is done with either an environment variable or yml configuration.
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
-You can use any of the Transport classes provided natively by SwiftMailer. There are also countless PHP libraries offering custom Transports to integrate with third party mailing service:
-- read the [SwiftMailer Transport Types documentation](https://swiftmailer.symfony.com/docs/sending.html#transport-types) for a full list of native Transport
-- search [Packagist for SwiftMailer Transport](https://packagist.org/?query=SwiftMailer+Transport) to discover additional third party integrations
+Alternatively you can provide a different `DSN` to select any of the Transport classes provided natively by `symfony/mailer` or other compatible third-party transports. For more information and to see what other transports are available see the [symfony/mailer transport types](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport).
```
```suggestion
Alternatively you can provide a different `DSN` to select any of the Transport classes provided natively by `symfony/mailer` or other compatible third-party transports. For more information and to see what other transports are available see the [symfony/mailer transport types](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport).

[hint]
The format for the DSN is exactly as defined in the symfony docs linked above. Some common examples are listed below.
[/hint]
```
Just in case anyone is confused - lets make it super clear that the symfony docs show what the dsn should look like - we're just passing it along.

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
-After:
-  - '#emailconfig'
+Name: mailer-dsn-config
+After: '*'
```
Does this need to be after everything?

emteknetnz replied:
I think it's safest to. This means it will always have precedence.

Technically it could possibly be `After: 'mailer-dsn-env'` - relates to https://github.com/silverstripe/silverstripe-framework/pull/10494/files

However
- I'm not 100% sure if 'mailer-dsn-env' yml block will always exist, as there's an `only` conditional in there which checks to see the `MAILER_DSN` env variable exists.
- we might need to add in another yml block that goes after 'mailer-dsn-env' at a later date, which would then take precedence over project config

There's a different argument to be made that the env variable should have precedence over any project yml (it's a pretty strong argument, I may have got it the wrong way around when I first though about this).

Probably makes sense to leave that argument for the actual framwork PR though

Should leave this unresolved for now

ChloeSartorelli replied:
I agree that this should be changed so that the .env config takes precedence (see https://github.com/silverstripe/silverstripe-framework/pull/10494/files#r985306124). Please update the docs to match.

emteknetnz replied:
Have updated docs, note: I've have kept the `MAILER_DSN` config in yml, rather than overriding in php.  This is because the service is created via new injector factory yml config. This also allows more flexibility in the case there's a misconfigured env variable on a prod server that needs to be urgently changed by developers who can't quickly change an env variable (though I haven't mentioned that explicitly in the docs).

I've updated the docs accordingly, specifically mentioning that project config must be exactly naming `mailer-dsn-project-config` to allow the env variable to retain precedence.

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
-      AuthMode: [ setAuthMode, ['login'] ]
+  Symfony\Component\Mailer\Transport\TransportInterface:
+    constructor:
+      dsn: '<my-dsn>'
```
The old docs had examples here for authentication - does symfony mailer simply never need authentication (I would find this highly suspect)? If it does, how do we pass that through? Is that all part of the dsn?
If so we should mention that fact above as well.

emteknetnz replied:
See below, I provided an example for an SMTP DSN string

ChloeSartorelli replied:
In that case:
> we should mention that fact above as well.

ChloeSartorelli replied:
Actually nevermind, the example below is close enough to here that it's not needed.

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -6,101 +6,78 @@ icon: envelope-open
 ```
 
-Note the usage of backticks to designate environment variables for the credentials - ensure you set these in your `.env` file or in your webserver configuration.
+The precedence for configuration, from highest to lowest:
````
```suggestion
The configuration priority order is as follows, from highest to lowest:
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
+### Common DSN strings
 
-### Mailer Configuration for dev environments
+To configure SMTP
```
```suggestion
#### To configure SMTP {#dsn-smtp}
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -6,101 +6,78 @@ icon: envelope-open
+```
 
-You can suppress all emails by using the [`Swift_Transport_NullTransport`](https://github.com/swiftmailer/swiftmailer/blob/master/lib/classes/Swift/Transport/NullTransport.php).
+To configure a different sendmail binary and command
````
```suggestion
#### To configure a different sendmail binary and command {#dsn-sendmail}
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,78 @@ icon: envelope-open
+Email configuration is done using Symfony's `[DSN](https://symfony.com/doc/current/mailer.html#transport-setup) configuration string which is used to select which transport is used and any required configuration such as username and password. In Silverstripe, this is done with either an environment variable or yml configuration.
 
-It's highly recommended you upgrade to a more robust transport for additional security. The Sendmail transport is the most common one. The `sendmail` binary is widely available across most Linux/Unix servers.
+The `Sendmail` transport is the most common one and is used by default in Silverstripe. The `sendmail` binary is widely available across most Linux/Unix servers. By default the sendmail command used is `/usr/sbin/sendmail -bs`.
```
```suggestion
The `Sendmail` transport is the most common one and is used by default in Silverstripe. The `sendmail` binary is widely available across most Linux/Unix servers. By default the sendmail command used is `/usr/sbin/sendmail -bs`, but this [can be configured](#dsn-sendmail) as part of the `DSN`.
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -6,101 +6,78 @@ icon: envelope-open
 ```
 
-If you're using MailCatcher, or a similar tool, you can tell `Swift_SendmailTransport` to use a different binary.
+To suppress all emails
````
```suggestion
#### To suppress all emails {#dsn-null}
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -6,101 +6,78 @@ icon: envelope-open
+MAILER_DSN="null://default"
+```
+
+Read more about other [DSN strings available](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport)
````
```suggestion
Read more about other available DSN strings in [the symfony documentation](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport)
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -110,11 +87,10 @@ Using the code snippet above also tests that the ability to set the "from" addre
 use SilverStripe\Control\Email\Email;
 
-$email = new Email($from, $to, $subject, $body);
+$email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');
```
I prefer the old example - most of the time people _will_ already have these in variables, so it's easy to just copy paste that into their projects, and the variables are self explanatory anyway.

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -124,44 +100,57 @@ By default, emails are sent in both HTML and Plaintext format. A plaintext repre
 from the system by stripping HTML markup, or transforming it where possible (e.g. `<strong>text</strong>` is converted 
 to `*text*`).
 
+```php
+use SilverStripe\Control\Email\Email;
+
+$email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');
+$email->send();
+```
````
```suggestion
```
This example is superfluous - it's already shown in the "Testing how email works" section which isn't far above here. At most the paragraph above could say "By default, when using the `send()` method" at the start.

emteknetnz replied:
Removed

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -124,44 +100,57 @@ By default, emails are sent in both HTML and Plaintext format. A plaintext repre
+$email->send();
+```
+
+You can also specify separate plain text content so that it's not automatically generated from HTML
````
```suggestion
You can also specify plain text and HTML content separately if you don't want the plain text to be automatically generated from HTML
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -124,44 +100,57 @@ By default, emails are sent in both HTML and Plaintext format. A plaintext repre
-$email = new Email($from, $to, $subject, $body);
+use SilverStripe\Control\Email\Email;
+
+$email = new Email('from@mydomain.com', 'to@example.com', 'My subject');
```
```suggestion
$email = new Email($from, $to, $subject);
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -124,44 +100,57 @@ By default, emails are sent in both HTML and Plaintext format. A plaintext repre
 [info]
-The default HTML template for emails is named `GenericEmail` and is located in `vendor/silverstripe/framework/templates/SilverStripe/Email/`.
+The default HTML template for emails is `vendor/silverstripe/framework/templates/SilverStripe/Control/Email/Email.ss`.
 To customise this template, copy it to the `app/templates/Email/` folder or use `setHTMLTemplate` when you create the 
```
Have you tested this? I think it would have to be in the `app/templates/SilverStripe/Control/Email/` folder.
Note: I haven't tested it yet - this first review is just a first pass finding what I can visually first.

emteknetnz replied:
Have updated instructions to say copy it to `<project-root>/themes/<my-theme>/SilverStripe/Control/Email/Email.ss` - I just tested this works out of the box without needing to specify the location with `setHTMLTemplate()`

ChloeSartorelli replied:
Yeah `setHTMLTemplate()` is for setting a specific template (e.g. if you have different templates for different emails)

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -124,44 +100,57 @@ By default, emails are sent in both HTML and Plaintext format. A plaintext repre
 [info]
-The default HTML template for emails is named `GenericEmail` and is located in `vendor/silverstripe/framework/templates/SilverStripe/Email/`.
+The default HTML template for emails is `vendor/silverstripe/framework/templates/SilverStripe/Control/Email/Email.ss`.
 To customise this template, copy it to the `app/templates/Email/` folder or use `setHTMLTemplate` when you create the 
```
```suggestion
To customise this template, copy it to the `app/templates/Email/` folder, or use `setHTMLTemplate()` to use a template in a different location when you create the 
```

emteknetnz replied:
Updated slightly differently based on change above

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -124,44 +100,57 @@ By default, emails are sent in both HTML and Plaintext format. A plaintext repre
 
-
-```ss
+```html
````
Revert this change. `ss` is correct - docs.silverstripe.org handles it correctly.
https://docs.silverstripe.org/en/4/developer_guides/email/#templates

emteknetnz replied:
Reverted

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -124,44 +100,57 @@ By default, emails are sent in both HTML and Plaintext format. A plaintext repre
-    ->setFrom($from)
-    ->setTo($to)
-    ->setSubject($subject);
+    ->from('from@mydomain.com')
+    ->to('to@example.com')
+    ->subject('My subject');
```
```suggestion
    ->from($from)
    ->to($to)
    ->subject($subject);
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -183,7 +171,9 @@ to specify your own own plaintext version/template you can use `$email->setPlain
-$email = new SilverStripe\Control\Email\Email();
+use SilverStripe\Control\Email\Email;
+
+$email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');
```
```suggestion
$email = new Email($from, $to, $subject, $body);
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
````diff
@@ -193,16 +183,14 @@ $this->send();
-
-
-```yaml
+```yml
````
Do you know for a fact this will format correctly? The existing `yaml` is definitely formatting correctly so I'm hesitant to change this. Probably best to leave these alone for now - there's [an existing card](https://github.com/silverstripe/developer-docs/issues/11) to standardise this sort of thing so if we want to standardise `yaml` to `yml` we can do it as part of that card.
I've added standardisation of `yaml` or `yml` to the card's proposed changes.

emteknetnz replied:
Reverted

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -250,45 +238,15 @@ necessarily the same which should be used for return correspondence and should h
-SilverStripe\Core\Injector\Injector:
-  Swift_Transport:
-    class: Swift_NullTransport
+$email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');
```
```suggestion
$email = new Email($from, $to, $subject, $body);
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -250,45 +238,15 @@ necessarily the same which should be used for return correspondence and should h
-For further information on SwiftMailer, consult their docs: http://swiftmailer.org/docs/introduction.html
+## Advanced customisation
 
+Silverstripe Email is built on top of symfony/mailer for for advanced customisation information refer to the [symfony/mailer docs](https://symfony.com/doc/current/mailer.html)
```
```suggestion
Silverstripe Email is built on top of [symfony/mailer](https://github.com/symfony/mailer). For advanced customisation information refer to the [symfony/mailer docs](https://symfony.com/doc/current/mailer.html)
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -9,6 +9,7 @@ title: 5.0.0 (unreleased)
 - [API changes](#api-changes)
   - [Removed API (by module, alphabetically)](#api-removed)
   - [General changes](#api-general)
+  - [ORM](#api-email)
```
```suggestion
  - [Email](#api-email)
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -60,6 +61,35 @@ This is a major release and as a result there are a number of breaking API chang
+    - `SwiftMailer`
+    - `SwiftPlugin`
+  - Removed the following methods from `SilverStripe\Control\Email\Email`, in most instances these have been replaced with equivalent functionality:
+    - `addAttachment()`
+    - `addAttachmentFromData()`
+    - `hasPlainPart()`
+    - `invalidateBody()`
+    - `findPlainPart()`
+    - `generatePlainPartFromBody()`
+    - `getBCC()`
+    - `getCC()`
+    - `getFailedRecipients()`
+    - `getFrom()`
+    - `getReturnPath()`
+    - `getSender()`
+    - `getSwiftMessage()`
+    - `getTo()`
+    - `setSwiftMessage()`
+    - `setFailedRecipients()`
+    - `BaseURL()`
+    - `IsEmail()`
```
For any of these which do have a new replacement, mention it here.
e.g. "`getTo()` replaced with `to()`"

This makes it a lot easier for people to know what changes they need to make.

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -60,6 +61,35 @@ This is a major release and as a result there are a number of breaking API chang
+    - `setFailedRecipients()`
+    - `BaseURL()`
+    - `IsEmail()`
+  - Changed return types on the following methods in `SilverStripe\Control\Email\Email`
+    - `send()`
+    - `sendPlain()`
+  - Changed visibility on the following methods to `private` in `SilverStripe\Control\Email\Email`
+    - `mergeConfiguredEmails()`
```
This section is only for _removed_ API.
Mention these below, in the email-specific section.

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -100,6 +130,15 @@ This is a major release and as a result there are a number of breaking API chang
 
+### Email {#api-email}
+
+- Email was sent in CMS 4 using [SwiftMailer](https://swiftmailer.symfony.com/docs/introduction.html). In CMS 5 this has been replaced with [symfony/mailer](https://symfony.com/doc/current/mailer.html).
+- `symfony/mailer` is the currently maintained email package from Symfony, and it also a more flexible email system that allows easier integration with third-party email providers
```
```suggestion
- Email was sent in CMS 4 using [SwiftMailer](https://swiftmailer.symfony.com/docs/introduction.html), which has since [gone End Of Life](https://symfony.com/blog/the-end-of-swiftmailer). In CMS 5 this has been replaced with [symfony/mailer](https://symfony.com/doc/current/mailer.html). `symfony/mailer` is the currently maintained email package from Symfony. It's a more flexible email system that allows easier integration with third-party email providers.
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -100,6 +130,15 @@ This is a major release and as a result there are a number of breaking API chang
+
+- Email was sent in CMS 4 using [SwiftMailer](https://swiftmailer.symfony.com/docs/introduction.html). In CMS 5 this has been replaced with [symfony/mailer](https://symfony.com/doc/current/mailer.html).
+- `symfony/mailer` is the currently maintained email package from Symfony, and it also a more flexible email system that allows easier integration with third-party email providers
+- Previously in CMS 4 there was official support for both `SendmailTransport` that used the sendmail binary available on most webservers, as well as `MailTransport` which uses the native php function `mail()`.
+- Older Silverstripe project were configured to use `MailTransport` out of the box. This was changed to `SendmailTransport` newer projects created using `silverstripe/installer` `4.9.0` or greater. `MailTransport` was added in a `thirdparty` folder in CMS 4 to retain support after it was officially removed in later versions of `SwiftMailer`.
```
I don't think this background information is necessary - lets just mention what has changed from 4 to 5 rather than what changed from 4 to another version of 4
```suggestion
```

emteknetnz replied:
Remooved

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -100,6 +130,15 @@ This is a major release and as a result there are a number of breaking API chang
+- `symfony/mailer` is the currently maintained email package from Symfony, and it also a more flexible email system that allows easier integration with third-party email providers
+- Previously in CMS 4 there was official support for both `SendmailTransport` that used the sendmail binary available on most webservers, as well as `MailTransport` which uses the native php function `mail()`.
+- Older Silverstripe project were configured to use `MailTransport` out of the box. This was changed to `SendmailTransport` newer projects created using `silverstripe/installer` `4.9.0` or greater. `MailTransport` was added in a `thirdparty` folder in CMS 4 to retain support after it was officially removed in later versions of `SwiftMailer`.
+- MailTransport, which used the php native `mail()` function, is no longer officially supported in CMS 5. This is because Symfony considers `mail()` to be [insecure](https://github.com/swiftmailer/swiftmailer/issues/866#issuecomment-289291228).
```
```suggestion
- `MailTransport`, which used the php native `mail()` function, is no longer officially supported in CMS 5. This is because [Symfony considers `mail()` to be insecure](https://github.com/swiftmailer/swiftmailer/issues/866#issuecomment-289291228).
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -100,6 +130,15 @@ This is a major release and as a result there are a number of breaking API chang
+- Previously in CMS 4 there was official support for both `SendmailTransport` that used the sendmail binary available on most webservers, as well as `MailTransport` which uses the native php function `mail()`.
+- Older Silverstripe project were configured to use `MailTransport` out of the box. This was changed to `SendmailTransport` newer projects created using `silverstripe/installer` `4.9.0` or greater. `MailTransport` was added in a `thirdparty` folder in CMS 4 to retain support after it was officially removed in later versions of `SwiftMailer`.
+- MailTransport, which used the php native `mail()` function, is no longer officially supported in CMS 5. This is because Symfony considers `mail()` to be [insecure](https://github.com/swiftmailer/swiftmailer/issues/866#issuecomment-289291228).
+- If your site has a custom email configuration e.g. SMTP configuration, this will need to be updated as the configuration has changed from Silverstripe yml to a much more flexible and standardised [DSN string](https://symfony.com/doc/current/mailer.html#transport-setup).
```
```suggestion
- If your site has a custom email configuration e.g. SMTP configuration, this will need to be updated, as the configuration has changed from Silverstripe yml to a much more flexible and standardised [DSN string](https://symfony.com/doc/current/mailer.html#transport-setup). See [the email documentation](/developer_guides/email/) for more details.
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -146,6 +186,7 @@ This release includes a number of bug fixes to improve a broad range of areas. C
 ## Environment variable changes
 
 - `SS_MANIFESTCACHE` can no longer use the now removed symfony/cache 4.x "Simple" cache classes e.g. `Symfony\Component\Cache\Simple\PhpFilesCache`. Instead use the corresponding "Adapter" class e.g. `Symfony\Component\Cache\Adapter\PhpFilesAdapter`.
+- `APP_SMTP_USERNAME` and `APP_SMTP_PASSWORD` no longer have any affect. Use a `MAILER_DSN` environment variable instead to configure SMTP email
```
```suggestion
- `APP_SMTP_USERNAME` and `APP_SMTP_PASSWORD` have been removed. Use a `MAILER_DSN` environment variable instead to configure SMTP email (see [the email documentation](/developer_guides/email/) for more details).
```
We may as well use stronger language here.

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -80,7 +106,7 @@ This is a major release and as a result there are a number of breaking API chang
 
 - Removed deprecated class `SilverStripe\VersionedAdmin\Controllers\HistoryControllerFactory`
 
-### General changes {#api-general}
+### General changes
```
Why did you remove the anchor?

emteknetnz replied:
Reverted

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -60,6 +61,31 @@ This is a major release and as a result there are a number of breaking API chang
+    - `SwiftMailer`
+    - `SwiftPlugin`
+  - Removed the following methods from `SilverStripe\Control\Email\Email`, in most instances these have been replaced with equivalent functionality:
+    - `addAttachment()`
+    - `addAttachmentFromData()`
+    - `debug()`
+    - `hasPlainPart()`
+    - `invalidateBody()`
+    - `findPlainPart()`
+    - `generatePlainPartFromBody()`
+    - `getBCC()`
+    - `getCC()`
+    - `getFailedRecipients()`
+    - `getFrom()`
+    - `getReturnPath()`
+    - `getSender()`
+    - `getSwiftMessage()`
+    - `getTo()`
+    - `setSwiftMessage()`
+    - `setFailedRecipients()`
+    - `BaseURL()`
+    - `IsEmail()`
```
The change mentioned in this comment: https://github.com/silverstripe/developer-docs/pull/92#discussion_r983071152 has not been done yet.

emteknetnz replied:
> getTo() replaced with to()
This isn't accurate - to() means setTo()

Silverstripe Email::getTo() used to return an array of strings (I think, possibly a string, or maybe something else),   Silverstripe getTo() has been removed.  Email::getTo() was just a wrapper for SwiftMailer SimpleMimeEntity getTo()

However there is still a getTo() method on Email - Symfony Email::getTo() which returns an array of Symfony Address object

So ... I feel this is really too intricate a thing to put in a changelog.  I think saying "in most instances these have been replaced with equivalent functionality" is good enough here.  Realistically very few projects will be using calling methods like `getTo()` because generally they only care about `setTo()` because they just want to send an email.  Possibly it'll cause some unit tests in projects to fail, but that will be easy to fix.

Looking at this list, only addAttachment() and addAttachmentFromData() are relevant - however these shouldn't even be here as I had already aliased them to the new method names, similar to what I did with the `setTo()` -> `to()` alias - this means less likely to have silent regressions on upgrade

ChloeSartorelli replied:
Apologies, I got `get` and `set` mixed up - the `getTo()` method, along with a number of others here, _are not being removed from the class's public API_.  Sure we're not implementing these methods _directly_ in `Email` anymore, but from the developers POV the method is still there. I can still call `$email->getTo()`. Saying we've removed those methods is just going to confuse people, since the methods are still present and callable, and they don't actually have to change anything about how they use them.

For any methods we are keeping (either directly or via inheritance), please remove them from this list.
For any methods are are removing, but which have a replacement (if any) please add "replaced with `theMethod()`" so that developers know what method they should be calling instead.

If any of the methods we're keeping have changed in a substantial way (changed parameters or returned values) please note this in the email section below if you haven't done so already.

emteknetnz replied:
Updated

ChloeSartorelli commented on en/02_Developer_Guides/10_Email/index.md:
```diff
@@ -6,101 +6,81 @@ icon: envelope-open
 # Email
 
-Creating and sending email in Silverstripe CMS is done through the [Email](api:SilverStripe\Control\Email\Email) and [Mailer](api:SilverStripe\Control\Email\Mailer) classes. This document covers how to create an `Email` instance, customise it with a HTML template, then send it through a custom `Mailer`.
+Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email). This document covers how to create an `Email` instance and customise it with an HTML template.
```
```suggestion
Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email) class. This document covers how to create an `Email` instance and customise it with an HTML template.
```
as per https://github.com/silverstripe/developer-docs/pull/92#discussion_r983053035

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -100,6 +126,18 @@ This is a major release and as a result there are a number of breaking API chang
 
+### Email {#api-email}
+
+- Email was sent in CMS 4 using [SwiftMailer](https://swiftmailer.symfony.com/docs/introduction.html), which has since [gone End Of Life](https://symfony.com/blog/the-end-of-swiftmailer). In CMS 5 this has been replaced with [symfony/mailer](https://symfony.com/doc/current/mailer.html). `symfony/mailer` is the currently maintained email package from Symfony. It's a more flexible email system that ...
```
We probably need to mention in here somewhere that `Email` is no longer a subclass of `ViewableData`, so anything people want to use in templates must be returned from `getData()`.

If the `getData()` method isn't mentioned in the main email docs, it also should be mentioned there (probably in the same area that talks about replacing the default template).

emteknetnz replied:
I'll add a bullet point to mention the subclass thing. The main email docs already talks about using setData() - https://docs.silverstripe.org/en/4/developer_guides/email/#templates

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -100,6 +124,19 @@ This is a major release and as a result there are a number of breaking API chang
+- `MailTransport`, which used the php native `mail()` function, is no longer officially supported in CMS 5. This is because [Symfony considers `mail()` to be insecure](https://github.com/swiftmailer/swiftmailer/issues/866#issuecomment-289291228).
+- If your site has a custom email configuration e.g. SMTP configuration, this will need to be updated, as the configuration has changed from Silverstripe yml to a much more flexible and standardised [DSN string](https://symfony.com/doc/current/mailer.html#transport-setup). See [the email documentation](/developer_guides/email/) for more details.
+- The following API changes also were made:
+  - Changed return types on the following methods in `SilverStripe\Control\Email\Email`
```
```suggestion
  - Changed return types on the following methods in `SilverStripe\Control\Email\Email` which no longer return anything
```
We should include all the pertinent information here - if we don't say it doesn't return anything, we're effectively forcing people to go look at the source code just to know what's changed.

emteknetnz replied:
Updated

GuySartorelli reviewed (changes requested):
We should also mention the strict typing changes (whether explicitly or just as a "we have also added strict types to many parameters and return types in this class" I'm not fussed) so people don't get caught out by it and then complain it wasn't in the changelogs.

Changes to parameter names are also API changes and need to be mentioned here.

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -62,6 +63,29 @@ This is a major release and as a result there are a number of breaking API chang
+    - Replaced with `Email::getData()`
+      - `BaseURL()`
+      - `IsEmail()`
+    - Has equivalent functionality via `Email::getTextBody()`
+      - `hasPlainPart()`
+      - `findPlainPart()`
```
We discussed in https://github.com/silverstripe/silverstripe-framework/pull/10517#discussion_r984063507 that these were removed without equivalent functionality - which is what the deprecation messages say.

If you're now saying they _do_ have equivalent functionality, please create a new PR targetting `4` with an updated deprecation message. Otherwise, please indicate these as not having equivalent functionality.

ChloeSartorelli replied:
Looking at the code in https://github.com/symfony/mime/blob/6.1/Email.php the docs here are correct, so we need to fix those deprecations.

emteknetnz replied:
I've moved `hasPlainPart()` and `findPlainPart()` to under the `Not replaced with equivalent functionality` section.  While they're kind of sort of replaced with `getTextBody()`, it's not really the same thing

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -62,6 +63,29 @@ This is a major release and as a result there are a number of breaking API chang
+    - Replaced with `Email::getData()`
+      - `BaseURL()`
+      - `IsEmail()`
+    - Has equivalent functionality via `Email::getTextBody()`
```
Why is the above "replaced with" but this is "has equivalent functionality via"?
IMO they should all say "replaced with" as that's much clearer language.

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -62,6 +63,29 @@ This is a major release and as a result there are a number of breaking API chang
+      - `getFailedRecipients()`
+      - `getSwiftMessage()`
+      - `mergeConfiguredEmails()`
+      - `render()`
```
This wasn't deprecated in the `4` PR - probably we now need a new PR to deprecate that if it's being removed.

emteknetnz replied:
It was? https://github.com/silverstripe/silverstripe-framework/blob/4/src/Control/Email/Email.php#L886

ChloeSartorelli replied:
Note this comment was from 12 days ago -  a lot of changes were made since then. :p
You're correct that _now_ it has been deprecated.

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -102,6 +126,25 @@ This is a major release and as a result there are a number of breaking API chang
+- `MailTransport`, which used the php native `mail()` function, is no longer officially supported in CMS 5. This is because [Symfony considers `mail()` to be insecure](https://github.com/swiftmailer/swiftmailer/issues/866#issuecomment-289291228).
+- If your site has a custom email configuration e.g. SMTP configuration, this will need to be updated, as the configuration has changed from Silverstripe yml to a much more flexible and standardised [DSN string](https://symfony.com/doc/current/mailer.html#transport-setup). See [the email documentation](/developer_guides/email/) for more details.
+- The following return types where changed on the following methods in `SilverStripe\Control\Email\Email`
+  - Used to return a string, now returns an nullable Address object
```
```suggestion
  - Used to return a string, now returns an nullable [Address](https://symfony.com/doc/current/mailer.html#email-addresses) object
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -102,6 +126,25 @@ This is a major release and as a result there are a number of breaking API chang
+  - Used to return a string, now returns an nullable Address object
+    - `getReturnPath()`
+    - `getSender()`
+  - Used to return an array of strings, now returns an array of Address objects
```
```suggestion
  - Used to return an array of strings, now returns an array of `Address` objects
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -102,6 +126,25 @@ This is a major release and as a result there are a number of breaking API chang
+    - `getTo()`
+    - `getBCC()` - note - changed casing to `getBcc()`
+    - `getCC()` - note - changed casing to `getCc()`
+  - Used to return bool, now return void:
```
```suggestion
  - Used to return `bool`, now return `void`:
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -62,6 +63,29 @@ This is a major release and as a result there are a number of breaking API chang
+- Removed various API related to `SwiftMailer`
+  - Removed the following classes and interfaces
+    - `SilverStripe\Control\Email\Mailer`
+    - `SwiftMailer`
+    - `SwiftPlugin`
```
If these had namespaces, please put the FQCN here.

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -102,6 +126,25 @@ This is a major release and as a result there are a number of breaking API chang
+- In CMS 4 the `SilverStripe\Control\Email\Email` class subclassed `SilverStripe\View\ViewableData`. In CMS 5 it now subclasses `Symfony\Component\Mime\Email`.
+- `MailTransport`, which used the php native `mail()` function, is no longer officially supported in CMS 5. This is because [Symfony considers `mail()` to be insecure](https://github.com/swiftmailer/swiftmailer/issues/866#issuecomment-289291228).
+- If your site has a custom email configuration e.g. SMTP configuration, this will need to be updated, as the configuration has changed from Silverstripe yml to a much more flexible and standardised [DSN string](https://symfony.com/doc/current/mailer.html#transport-setup). See [the email documentation](/developer_guides/email/) for more details.
+- The following return types where changed on the following methods in `SilverStripe\Control\Email\Email`
```
```suggestion
- The following return types where changed on the following methods in [Email](api:SilverStripe\Control\Email\Email)
```

emteknetnz replied:
Updated

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -102,6 +126,25 @@ This is a major release and as a result there are a number of breaking API chang
+### Email {#api-email}
+
+- Email was sent in CMS 4 using [SwiftMailer](https://swiftmailer.symfony.com/docs/introduction.html), which has since [gone End Of Life](https://symfony.com/blog/the-end-of-swiftmailer). In CMS 5 this has been replaced with [symfony/mailer](https://symfony.com/doc/current/mailer.html). `symfony/mailer` is the currently maintained email package from Symfony. It's a more flexible email system that ...
+- In CMS 4 the `SilverStripe\Control\Email\Email` class subclassed `SilverStripe\View\ViewableData`. In CMS 5 it now subclasses `Symfony\Component\Mime\Email`.
```
```suggestion
- In CMS 4 the [Email](api:SilverStripe\Control\Email\Email) class subclassed [ViewableData](api:SilverStripe\View\ViewableData). In CMS 5 it now subclasses `Symfony\Component\Mime\Email`.
```

ChloeSartorelli replied:
This has not been done? Along with several other requested changes from the looks of it.

emteknetnz replied:
Rather than play peer review ping pong to update every Silverstripe `FQCN` to an api link, I think it'll be far more efficient to simply do this in one big go when we finalize the 5.0.0 changelog prior to releasing 5.0.0.

There's currently loads of other Silverstripe `FQCN` in the changelog unrelated to email that aren't api links

ChloeSartorelli replied:
It's not ping pong if you make the change :p But I take your point, I can live with it for now.

ChloeSartorelli commented on en/04_Changelogs/5.0.0.md:
```diff
@@ -102,6 +126,25 @@ This is a major release and as a result there are a number of breaking API chang
 - The constructor for `SilverStripe\Versioned\Versioned` now explicitly only accepts mode as a single argument.
 
+### Email {#api-email}
+
```
Please add a quick sentence here right near the top that says something completely clear like "The configuration for email has changed completely - please read the updated documentation carefully" (as per https://github.com/silverstripe/silverstripe-framework/pull/10494#discussion_r996567072)

emteknetnz replied:
Added

emteknetnz commented:
@GuySartorelli 
> We should also mention the strict typing changes (whether explicitly or just as a "we have also added strict types to many parameters and return types in this class" I'm not fussed) so people don't get caught out by it and then complain it wasn't in the changelogs.
> Changes to parameter names are also API changes and need to be mentioned here.

I've added a sentence for these two things

GuySartorelli reviewed (approved):
Looks good. I'll hold off merging until the framework PR is ready to merge just in case.
